book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+> for \\(j = 1\\). The final `[]` is delimited by D4 and D5 for \\(j = 1\\).
+> It does not contain an inner list item because D5/\\(j = 0\\) is not
+> asserted, so the outer list is empty.
+
+The state of lane \\(i\\) in the `data` signal is significant only while
+`valid` is asserted.
+
+> Note that the validity of the `last` lanes is thus *not* controlled by
+> which data lanes are active. This allows sequences to be terminated without
+> sending an element; this is in fact required in order to send empty
+> sequences, but also allows postponing the end-of-sequence marker until after
+> the last element (if complexity allows) as the previous example shows.
+
+It is illegal to assert `last` bit \\(j\\) without first terminating all
+\\(j' < j\\), if any data lanes were active since the previous assertion of
+`last` bit \\(j" = 0\\).
+
+> For example, the following is illegal because of the above:
+>
+> ```text
+>         .------------.
+>       0 | data:  '1' |
+>         | last:  0 0 |
+>         |------------|
+>       1 | data:  '2' |
+>         | last:  0 1 |
+>         |------------|
+> lane  2 | data:  '3' |
+> index   | last:  0 0 |
+>   |     |------------|
+>   v   3 | data:  '4' |
+>         | last:  1 0 |
+>         |------------|
+>       4 | data:  '5' |
+>         | last:  0 0 |
+>         |------------|
+>       5 | data:  '6' |
+>         | last:  1 1 |
+>         '------------'
+> ```
+>
+> An argument could be made that this encodes `[[1, 2], 3, 4], [[5, 6]]`, but
+> allowing this interpretation would make the stream significantly less
+> type-safe, and there would be no way to encode non-list elements before the
+> last inner list in the outer list without introducing a `start` signal as
+> well.
+>
+> An argument could also be made that this encodes `[[1, 2]], [[3, 4, 5, 6]]`,
+> because the `last` flag delimiting `[1, 2]` falls within the first outer list
+> and the `last` flag delimiting `[3, 4, 5, 6]` falls within the second. However,
+> This would unnecessarily complicate sink logic, make manual interpretation of
+> the data structure more context-sensitive, and so on.
+>
+> One could also argue that it encodes `[[1, 2]], [[5, 6]]`, because `3` and
+> `4` are not enclosed by any inner list, but then it makes more sense to just
+> deactivate the lanes.
+>
+> Ultimately, this representation is ambiguous or at least leads to confusion,
+> and is therefore illegal.
+
+*[\\(C < 8\\)]* All last bits for lanes 0 to \\(N - 2\\) inclusive must be
+driven low by the source, and may be ignored by the sink.
+
+> The above rule ultimately means that the `last` information is transferred on
+> a transfer basis instead of on element basis, similar to AXI4-stream. This
+> can significantly decrease decoding complexity, but only allows one innermost
+> sequence to be transferred per cycle regardless of \((N\\) and the length of
+> the sequence.
+
+*[\\(C < 4\\)]* It is illegal to assert `last` bit \\(j\\) without also
```

Comment:
While I agree that my attempt at mathematically identifying which `last` bit index I'm talking about turned out pretty poor, there is no "the `last` bit" because there can be multiple, I'm talking about the one with dimension j. I'll try to clarify the text.

> See below.

Huh?

Reply:
I misread this.